googletest/src/gtest.cc

```diff
@@ -800,41 +807,50 @@ std::string UnitTestImpl::CurrentOsStackTraceExceptTop(int skip_count) {
+#elif GTEST_OS_WINDOWS 
+  static const long long _Freq = _Query_perf_frequency(); // doesn't change after system boot
+  const long long _Ctr = _Query_perf_counter();
+  const long long _Whole = (_Ctr / _Freq);
+  const long long _Part = (_Ctr % _Freq);
+  return static_cast<TimeInMillis>(_Whole * 1000 + (_Part * 1000) / _Freq);
+#elif GTEST_OS_LINUX
+  timespec now;
+  clock_gettime(CLOCK_MONOTONIC, &now);
+  return static_cast<TimeInMillis>(now.tv_sec) * 1000 + static_cast<TimeInMillis>(now.tv_nsec / 1000000);
+#elif GTEST_OS_MAC
+  static bool initTimeBase = true;
+  static mach_timebase_info_data_t timebaseInfo;
+  if (initTimeBase)
+  {
+      initTimeBase = false;
+      mach_timebase_info(&timebaseInfo);
+  }
+  const unsigned long long now = mach_absolute_time();
+  return static_cast<TimeInMillis>((now / 1000000) * timebaseInfo.numer / timebaseInfo.denom);
+#else
+  // fall back to non monotonic time
+  return GetTimeInMillis();
+#endif
+
+}
+
 // Returns the current time in milliseconds.
 TimeInMillis GetTimeInMillis() {
-#if GTEST_OS_WINDOWS_MOBILE || defined(__BORLANDC__)
-  // Difference between 1970-01-01 and 1601-01-01 in milliseconds.
+#if (GTEST_OS_WINDOWS || defined(__BORLANDC__)) && !GTEST_HAS_GETTIMEOFDAY_
+  // 100ns ticks between 1/1/1601 00:00 UTC and 1/1/1970 00:00 UTC
   // http://analogous.blogspot.com/2005/04/epoch.html
-  const TimeInMillis kJavaEpochToWinFileTimeDelta =
-    static_cast<TimeInMillis>(116444736UL) * 100000UL;
-  const DWORD kTenthMicrosInMilliSecond = 10000;
-
-  SYSTEMTIME now_systime;
-  FILETIME now_filetime;
-  ULARGE_INTEGER now_int64;
-  // TODO(kenton@google.com): Shouldn't this just use
-  //   GetSystemTimeAsFileTime()?
-  GetSystemTime(&now_systime);
-  if (SystemTimeToFileTime(&now_systime, &now_filetime)) {
-    now_int64.LowPart = now_filetime.dwLowDateTime;
-    now_int64.HighPart = now_filetime.dwHighDateTime;
-    now_int64.QuadPart = (now_int64.QuadPart / kTenthMicrosInMilliSecond) -
-      kJavaEpochToWinFileTimeDelta;
-    return now_int64.QuadPart;
-  }
-  return 0;
-#elif GTEST_OS_WINDOWS && !GTEST_HAS_GETTIMEOFDAY_
-  __timeb64 now;
-
-  // MSVC 8 deprecates _ftime64(), so we want to suppress warning 4996
-  // (deprecated function) there.
-  // TODO(kenton@google.com): Use GetTickCount()?  Or use
-  //   SystemTimeToFileTime()
-  GTEST_DISABLE_MSC_WARNINGS_PUSH_(4996)
-  _ftime64(&now);
-  GTEST_DISABLE_MSC_WARNINGS_POP_()
-
-  return static_cast<TimeInMillis>(now.time) * 1000 + now.millitm;
+  const unsigned long long epochTime = 0x19DB1DED53E8000ULL; 
+  FILETIME now;
+  GetSystemTimeAsFileTime(&now);
+  unsigned long long fileTimeTicks = (static_cast<unsigned long long>(now.dwHighDateTime) << (sizeof(DWORD) * 8)) +
+                                     static_cast<unsigned long long>(now.dwLowDateTime) - epochTime;
+  return static_cast<TimeInMillis>(fileTimeTicks / 10000);
 #elif GTEST_HAS_GETTIMEOFDAY_
   struct timeval now;
   gettimeofday(&now, NULL);
```

Comment:
This is not monotonic time.

Reply:
you are correct.  The routines in GetTimeInMillis() return the time since POSIX time of 1/1/1970 00:00 UTC.  GetTimeInMillis is used in code to set start time with regards to wall clock.

The routines in GetMonotonicTimeInMillis compute monotonic clocks for elapsed time computations

Reply:
Some more comments.
I'll try to spin a some tests with this next week. Thanks.